File: api/src/test/kotlin/no/nav/pto_admin/controller/TilgangOppslagControllerTest.java
package no.nav.pto_admin.controller;

import no.nav.common.types.identer.EnhetId;
import no.nav.common.types.identer.NavIdent;
import no.nav.common.types.identer.NorskIdent;
import no.nav.pto_admin.service.AuthService;
import no.nav.pto_admin.service.TilgangOppslagService;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.test.context.junit4.SpringRunner;
import org.springframework.test.web.servlet.MockMvc;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.content;

@RunWith(SpringRunner.class)
@WebMvcTest(controllers = {TilgangOppslagController.class})
public class TilgangOppslagControllerTest {

    @Autowired
    private MockMvc mockMvc;

    @MockBean
    private AuthService authService;

    @MockBean
    private TilgangOppslagService tilgangOppslagService;

    @Test
    public void harTilgangTilEnhet__skal_sjekke_tilgang_til_pto_admin() throws Exception {
        mockMvc.perform(
                get("/api/tilgang/enhet")
                        .queryParam("navIdent", "Z1234")
                        .queryParam("enhetId", "1234")
        );

        verify(authService, times(1)).sjekkTilgangTilPtoAdmin();
    }

    @Test
    public void harTilgangTilEnhet__skal_sjekke_tilgang_med_parameter() throws Exception {
        mockMvc.perform(
                get("/api/tilgang/enhet")
                        .queryParam("navIdent", "Z1234")
                        .queryParam("enhetId", "1234")
        );

        verify(tilgangOppslagService, times(1))
                .harTilgangTilEnhet(NavIdent.of("Z1234"), EnhetId.of("1234"));
    }

    @Test
    public void harTilgangTilEnhet__skal_returnere_har_tilgang() throws Exception {
        when(tilgangOppslagService.harTilgangTilEnhet(any(), any())).thenReturn(true);

        mockMvc.perform(
                get("/api/tilgang/enhet")
                        .queryParam("navIdent", "Z1234")
                        .queryParam("enhetId", "1234")
        ).andExpect(content().json("{ \"harTilgang\": true  }"));
    }

    // =========================

    @Test
    public void harSkrivetilgang__skal_sjekke_tilgang_til_pto_admin() throws Exception {
        mockMvc.perform(
                get("/api/tilgang/skriv")
                        .queryParam("navIdent", "Z1234")
                        .queryParam("norskIdent", "1234567")
        );

        verify(authService, times(1)).sjekkTilgangTilPtoAdmin();
    }

    @Test
    public void harSkrivetilgang__skal_sjekke_tilgang_med_parameter() throws Exception {
        mockMvc.perform(
                get("/api/tilgang/skriv")
                        .queryParam("navIdent", "Z1234")
                        .queryParam("norskIdent", "1234567")
        );

        verify(tilgangOppslagService, times(1))
                .harSkrivetilgang(NavIdent.of("Z1234"), NorskIdent.of("1234567"));
    }

    @Test
    public void harSkrivetilgang__skal_returnere_har_tilgang() throws Exception {
        when(tilgangOppslagService.harSkrivetilgang(any(), any())).thenReturn(true);

        mockMvc.perform(
                get("/api/tilgang/skriv")
                        .queryParam("navIdent", "Z1234")
                        .queryParam("norskIdent", "1234567")
        ).andExpect(content().json("{ \"harTilgang\": true  }"));
    }

    // =========================

    @Test
    public void harLesetilgang__skal_sjekke_tilgang_til_pto_admin() throws Exception {
        mockMvc.perform(
                get("/api/tilgang/les")
                        .queryParam("navIdent", "Z1234")
                        .queryParam("norskIdent", "1234567")
        );

        verify(authService, times(1)).sjekkTilgangTilPtoAdmin();
    }

    @Test
    public void harLesetilgang__skal_sjekke_tilgang_med_parameter() throws Exception {
        mockMvc.perform(
                get("/api/tilgang/les")
                        .queryParam("navIdent", "Z1234")
                        .queryParam("norskIdent", "1234567")
        );

        verify(tilgangOppslagService, times(1))
                .harLesetilgang(NavIdent.of("Z1234"), NorskIdent.of("1234567"));
    }

    @Test
    public void harLesetilgang__skal_returnere_har_tilgang() throws Exception {
        when(tilgangOppslagService.harLesetilgang(any(), any())).thenReturn(true);

        mockMvc.perform(
                get("/api/tilgang/les")
                        .queryParam("navIdent", "Z1234")
                        .queryParam("norskIdent", "1234567")
        ).andExpect(content().json("{ \"harTilgang\": true  }"));
    }

    // =========================

    @Test
    public void harTilgangTilKode6__skal_sjekke_tilgang_til_pto_admin() throws Exception {
        mockMvc.perform(
                get("/api/tilgang/kode6")
                        .queryParam("navIdent", "Z1234")
        );

        verify(authService, times(1)).sjekkTilgangTilPtoAdmin();
    }

    @Test
    public void harTilgangTilKode6__skal_sjekke_tilgang_med_parameter() throws Exception {
        mockMvc.perform(
                get("/api/tilgang/kode6")
                        .queryParam("navIdent", "Z1234")
        );

        verify(tilgangOppslagService, times(1))
                .harTilgangTilKode6(NavIdent.of("Z1234"));
    }

    @Test
    public void harTilgangTilKode6__skal_returnere_har_tilgang() throws Exception {
        when(tilgangOppslagService.harTilgangTilKode6(any())).thenReturn(true);

        mockMvc.perform(
                get("/api/tilgang/kode6")
                        .queryParam("navIdent", "Z1234")
        ).andExpect(content().json("{ \"harTilgang\": true  }"));
    }

    // =========================

    @Test
    public void harTilgangTilKode7__skal_sjekke_tilgang_til_pto_admin() throws Exception {
        mockMvc.perform(
                get("/api/tilgang/kode7")
                        .queryParam("navIdent", "Z1234")
        );

        verify(authService, times(1)).sjekkTilgangTilPtoAdmin();
    }

    @Test
    public void harTilgangTilKode7__skal_sjekke_tilgang_med_parameter() throws Exception {
        mockMvc.perform(
                get("/api/tilgang/kode7")
                        .queryParam("navIdent", "Z1234")
        );

        verify(tilgangOppslagService, times(1))
                .harTilgangTilKode7(NavIdent.of("Z1234"));
    }

    @Test
    public void harTilgangTilKode7__skal_returnere_har_tilgang() throws Exception {
        when(tilgangOppslagService.harTilgangTilKode7(any())).thenReturn(true);

        mockMvc.perform(
                get("/api/tilgang/kode7")
                        .queryParam("navIdent", "Z1234")
        ).andExpect(content().json("{ \"harTilgang\": true  }"));
    }

    // =========================

    @Test
    public void harTilgangTilSkjermetPerson__skal_sjekke_tilgang_til_pto_admin() throws Exception {
        mockMvc.perform(
                get("/api/tilgang/skjermet")
                        .queryParam("navIdent", "Z1234")
        );

        verify(authService, times(1)).sjekkTilgangTilPtoAdmin();
    }

    @Test
    public void harTilgangTilSkjermetPerson__skal_sjekke_tilgang_med_parameter() throws Exception {
        mockMvc.perform(
                get("/api/tilgang/skjermet")
                        .queryParam("navIdent", "Z1234")
        );

        verify(tilgangOppslagService, times(1))
                .harTilgangTilSkjermetPerson(NavIdent.of("Z1234"));
    }

    @Test
    public void harTilgangTilSkjermetPerson__skal_returnere_har_tilgang() throws Exception {
        when(tilgangOppslagService.harTilgangTilSkjermetPerson(any())).thenReturn(true);

        mockMvc.perform(
                get("/api/tilgang/skjermet")
                        .queryParam("navIdent", "Z1234")
        ).andExpect(content().json("{ \"harTilgang\": true  }"));
    }

}

File: api/src/test/kotlin/no/nav/pto_admin/controller/TilgangOppslagControllerTest.kt
package no.nav.pto_admin.controller

import no.nav.common.types.identer.EnhetId
import no.nav.common.types.identer.NorskIdent
import no.nav.pto_admin.config.ApplicationTestConfig
import no.nav.pto_admin.config.SetupLocalEnvironment
import no.nav.pto_admin.service.TilgangOppslagService
import org.junit.jupiter.api.Test
import org.mockito.kotlin.whenever
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.reactive.WebFluxTest
import org.springframework.boot.test.mock.mockito.MockBean
import org.springframework.security.test.web.reactive.server.SecurityMockServerConfigurers
import org.springframework.test.context.ContextConfiguration
import org.springframework.test.web.reactive.server.WebTestClient
import java.util.*

@ContextConfiguration(classes = [ApplicationTestConfig::class])
@WebFluxTest(
    controllers = [TilgangOppslagController::class]
)
class TilgangOppslagControllerTest {

    init {
        SetupLocalEnvironment.setup()
    }

    @Autowired
    private lateinit var webClient: WebTestClient

    @MockBean
    private lateinit var tilgangOppslagService: TilgangOppslagService
    private val navAnsattAzureId: UUID = UUID.randomUUID()

    @Test
    fun harTilgangTilEnhet__skal_returnere_har_tilgang_false() {
        whenever(tilgangOppslagService.harTilgangTilEnhet(navAnsattAzureId, EnhetId("1234"))).thenReturn(false)

        loggedInWebClient().get().uri("/api/tilgang/enhet?navAnsattAzureId=${navAnsattAzureId}&enhetId=1234")
            .exchange()
            .expectStatus().isOk
            .expectBody(String::class.java)
            .isEqualTo("""{"harTilgang":false}""")
    }

    @Test
    fun harTilgangTilEnhet__skal_returnere_har_tilgang_true() {
        whenever(tilgangOppslagService.harTilgangTilEnhet(navAnsattAzureId, EnhetId("1234"))).thenReturn(true)

        loggedInWebClient().get().uri("/api/tilgang/enhet?navAnsattAzureId=$navAnsattAzureId&enhetId=1234")
            .exchange()
            .expectStatus().isOk
            .expectBody(String::class.java)
            .isEqualTo("""{"harTilgang":true}""")
    }

    @Test
    fun harSkrivetilgang__skal_returnere_har_tilgang_false() {
        whenever(tilgangOppslagService.harSkrivetilgang(navAnsattAzureId, NorskIdent("1234567"))).thenReturn(false)

        loggedInWebClient().get().uri("/api/tilgang/skriv?navAnsattAzureId=$navAnsattAzureId&norskIdent=1234567")
            .exchange()
            .expectStatus().isOk
            .expectBody(String::class.java)
            .isEqualTo("""{"harTilgang":false}""")
    }

    @Test
    fun harSkrivetilgang__skal_returnere_har_tilgang_true() {
        whenever(tilgangOppslagService.harSkrivetilgang(navAnsattAzureId, NorskIdent("1234567"))).thenReturn(true)

        loggedInWebClient().get().uri("/api/tilgang/skriv?navAnsattAzureId=$navAnsattAzureId&norskIdent=1234567")
            .exchange()
            .expectStatus().isOk
            .expectBody(String::class.java)
            .isEqualTo("""{"harTilgang":true}""")
    }

    @Test
    fun harLesetilgang__skal_returnere_har_tilgang_false() {
        whenever(tilgangOppslagService.harLesetilgang(navAnsattAzureId, NorskIdent("1234567"))).thenReturn(false)

        loggedInWebClient().get().uri("/api/tilgang/les?navAnsattAzureId=$navAnsattAzureId&norskIdent=1234567")
            .exchange()
            .expectStatus().isOk
            .expectBody(String::class.java)
            .isEqualTo("""{"harTilgang":false}""")
    }

    @Test
    fun harLesetilgang__skal_returnere_har_tilgang_true() {
        whenever(tilgangOppslagService.harLesetilgang(navAnsattAzureId, NorskIdent("1234567"))).thenReturn(true)

        loggedInWebClient().get().uri("/api/tilgang/les?navAnsattAzureId=$navAnsattAzureId&norskIdent=1234567")
            .exchange()
            .expectStatus().isOk
            .expectBody(String::class.java)
            .isEqualTo("""{"harTilgang":true}""")
    }

    @Test
    fun harTilgangTilKode6__skal_returnere_har_tilgang_false() {
        whenever(tilgangOppslagService.harTilgangTilKode6(navAnsattAzureId)).thenReturn(false)

        loggedInWebClient().get().uri("/api/tilgang/kode6?navAnsattAzureId=$navAnsattAzureId")
            .exchange()
            .expectStatus().isOk
            .expectBody(String::class.java)
            .isEqualTo("""{"harTilgang":false}""")
    }

    @Test
    fun harTilgangTilKode6__skal_returnere_har_tilgang_true() {
        whenever(tilgangOppslagService.harTilgangTilKode6(navAnsattAzureId)).thenReturn(true)

        loggedInWebClient().get().uri("/api/tilgang/kode6?navAnsattAzureId=$navAnsattAzureId")
            .exchange()
            .expectStatus().isOk
            .expectBody(String::class.java)
            .isEqualTo("""{"harTilgang":true}""")
    }

    @Test
    fun harTilgangTilKode7__skal_returnere_har_tilgang_false() {
        whenever(tilgangOppslagService.harTilgangTilKode7(navAnsattAzureId)).thenReturn(false)

        loggedInWebClient().get().uri("/api/tilgang/kode7?navAnsattAzureId=$navAnsattAzureId")
            .exchange()
            .expectStatus().isOk
            .expectBody(String::class.java)
            .isEqualTo("""{"harTilgang":false}""")
    }

    @Test
    fun harTilgangTilKode7__skal_returnere_har_tilgang_true() {
        whenever(tilgangOppslagService.harTilgangTilKode7(navAnsattAzureId)).thenReturn(true)

        loggedInWebClient().get().uri("/api/tilgang/kode7?navAnsattAzureId=$navAnsattAzureId")
            .exchange()
            .expectStatus().isOk
            .expectBody(String::class.java)
            .isEqualTo("""{"harTilgang":true}""")
    }

    @Test
    fun harTilgangTilSkjermetPerson__skal_returnere_har_tilgang_false() {
        whenever(tilgangOppslagService.harTilgangTilSkjermetPerson(navAnsattAzureId)).thenReturn(false)

        loggedInWebClient().get().uri("/api/tilgang/skjermet?navAnsattAzureId=$navAnsattAzureId")
            .exchange()
            .expectStatus().isOk
            .expectBody(String::class.java)
            .isEqualTo("""{"harTilgang":false}""")
    }

    @Test
    fun harTilgangTilSkjermetPerson__skal_returnere_har_tilgang_true() {
        whenever(tilgangOppslagService.harTilgangTilSkjermetPerson(navAnsattAzureId)).thenReturn(true)

        loggedInWebClient().get().uri("/api/tilgang/skjermet?navAnsattAzureId=$navAnsattAzureId")
            .exchange()
            .expectStatus().isOk
            .expectBody(String::class.java)
            .isEqualTo("""{"harTilgang":true}""")
    }

    fun loggedInWebClient(): WebTestClient {
        return webClient.mutateWith(SecurityMockServerConfigurers.mockOidcLogin())
    }
}
